### PR TITLE
feat(discover2): Store discover state as an LZ string

### PR DESCRIPTION
**Pull request by dashed**
## TODO

- [ ] update tests
- [ ] remove stats generation commit
- [x] improve interop with saved queries

-----

The following are the resulting character savings between the new resulting query string in comparison to the original: 

```
query: All Events - savings of 30 characters
query: Project Summary - savings of -20 characters
query: Errors - savings of 6 characters
query: Errors by URL - savings of -25 characters
query: Errors by User - savings of -16 characters
query: CSP - savings of 18 characters
query: CSP Report by Directive - savings of -14 characters
query: CSP Report by Blocked URI - savings of -34 characters
query: CSP Report by User - savings of -20 characters
query: Transactions - savings of -11 characters
query: Transactions by User - savings of -5 characters
query: Transactions by Region - savings of -36 characters
```

-----

Closes SEN-1036

### CODE REVIEW

File: src/sentry/static/sentry/app/views/eventsV2/eventView.tsx [hunk 1]
```diff
@@ -27,32 +34,57 @@ type Field = {
...
+      title = aliases[index];
     }
     parsed.push({field, title});
   });
 
   return parsed;
 };
 
+type DiscoverState = {
+  fields: string[];
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: Should the name of the query view go into the state as well? What about the datetime, query text and project selections? Those are all part of the saved query too.

Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: I can put some query strings into the state to be compressed; but not all. 

The query strings here are set by the component that controls the project selection, env selection, and the time span: https://github.com/getsentry/sentry/blob/05a1de9d32fefa5676c8065edd73f044ce9aa9ac/src/sentry/static/sentry/app/views/eventsV2/eventView.tsx#L366-L387

Comment (markstory) [reply src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: I forgot about global selection state needing to be managed elsewhere.

File: src/sentry/static/sentry/app/views/eventsV2/eventView.tsx [hunk 2]
```diff
@@ -217,18 +206,49 @@ class EventView {
...
-      name: decodeScalar(location.query.name),
-      fields: decodeFields(location),
-      sorts: decodeSorts(location),
-      tags: decodeTags(location),
-      query: decodeQuery(location),
+    const newLength = location.search.length;
+
+    const discoverState = intoDiscoverState(location);
+
+    const event_view = new EventView({
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: js vars are usually camelBack. 🏌

File: src/sentry/static/sentry/app/views/eventsV2/eventView.tsx [hunk 3]
```diff
@@ -217,18 +206,49 @@ class EventView {
...
+
+    const oldLength = oldQueryString.length;
+
+    console.log(
+      'query:',
+      event_view.name,
+      '- savings of',
+      oldLength - newLength,
+      'characters'
+    );
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: oopsie.

File: src/sentry/static/sentry/app/views/eventsV2/eventView.tsx [hunk 4]
```diff
@@ -27,32 +35,62 @@ type Field = {
...
-  }
-
-  const fields: string[] = isString(query.field) ? [query.field] : query.field;
-  const aliases: string[] = Array.isArray(query.alias)
-    ? query.alias
-    : isString(query.alias)
-    ? [query.alias]
-    : [];
+const decodeFields = (fields: string[], aliases: string[]): Field[] => {
+  // invariant: fields and aliases should have equal array lengths
```
[the remark this answers is not in the document]
Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: Ah interesting. I'll refactor with this assumption.

**Pull request closed without merging**